mlir/lib/QEC/Transforms/TLayerReduction.cpp

```diff
@@ -0,0 +1,277 @@
+            return std::pair(false, nullptr);
+        }
+
+        assert(lhsOp != rhsOp && "lshOp and rhsOp should not be equal");
+        assert(lhsOp->isBeforeInBlock(rhsOp) && "lhsOp should be before rhsOp");
+
+        // Reaching in-qubit values of `rhsOp` at the program point of `lhsOp`.
+        std::vector<Value> rhsOpInQubitsFromLhsOp = getInQubitReachingValuesAt(rhsOp, lhsOp);
+        if (llvm::any_of(rhsOpInQubitsFromLhsOp, [](Value qubit) { return qubit == nullptr; })) {
+            // Missing reaching value (intervening non-PPR); commutation fails.
+            // TODO: Handle non-PPR not directly dominated by `lhsOp`.
+            return std::pair(false, nullptr);
+        }
+
+        // Normalize to Pauli strings
+        auto normalizedOps =
+            normalizePPROps(lhsOp, rhsOp, lhsOp.getInQubits(), rhsOpInQubitsFromLhsOp);
+
+        if (!normalizedOps.first.commutes(normalizedOps.second)) {
+            return std::pair(false, nullptr);
+        }
+
+        // Equal normalized Pauli strings => merge candidate
+        auto canMerge = equal(normalizedOps.first, normalizedOps.second);
+        if (canMerge) {
+            mergeOp = lhsOp;
+        }
+    }
+
+    return std::pair(true, mergeOp);
+}
+
+void moveOpToLayer(QECOpInterface rhsOp, QECLayer &rhsLayer, QECOpInterface mergeOp,
+                   QECLayer &lhsLayer, IRRewriter &writer)
+{
+    //    lhsLayer   :  rhsLayer
+    //    ┌───────┐  :  ┌───────┐
+    //   ─┤ lhsOp ├──:──┤ rhsOp ├─
+    //    └───────┘  :  └───────┘
+
+    // Move `rhsOp` into `lhsLayer` and remap its operands to values visible at `lhsOp`.
+
+    auto lhsOp = lhsLayer.getOps().back();
+
+    if (mergeOp) {
+        lhsOp = mergeOp;
+    }
+
+    auto newOp = rhsOp.clone();
+
+    // Map operands via values at `lhsOp`; if unused by `lhsOp`, use the value
+    // available before `lhsOp`.
+    std::vector<Value> newOperands;
+    IRMapping lhsInOutQubits;
+    lhsInOutQubits.map(lhsOp.getInQubits(), lhsOp.getOutQubits());
+
+    for (auto opnd : rhsOp.getInQubits()) {
+        auto derivedOpnd = getReachingValueAt(opnd, lhsOp);
+        assert(derivedOpnd != nullptr && "Qubit should not be nullptr");
+
+        if (auto out = lhsInOutQubits.lookupOrNull(derivedOpnd)) {
+            derivedOpnd = out;
+        }
+        newOperands.emplace_back(derivedOpnd);
+    }
+
+    newOp->setOperands(newOperands);
+
+    // Replace uses of newOp's in-qubits with its out-qubits.
+    for (auto [inNewOp, outNewOp] : llvm::zip(newOp.getInQubits(), newOp.getOutQubits())) {
+        writer.replaceAllUsesExcept(inNewOp, outNewOp, newOp);
+    }
+
+    writer.setInsertionPointAfter(lhsOp);
+    writer.insert(newOp);
+    lhsLayer.insertToLayer(newOp);
+
+    rhsLayer.eraseOp(rhsOp);
+    // Rewire users of the erased `rhsOp` to its original operands.
+    writer.replaceAllUsesWith(rhsOp->getResults(), rhsOp->getOperands());
+    writer.eraseOp(rhsOp);
```

Comment:
You can just using `replaceOp`
```suggestion
    writer.replaceOp(rhsOp, rhsOp->getOperands());
```